datafusion/src/logical_plan/builder.rs

```diff
@@ -1090,6 +1094,101 @@ pub(crate) fn expand_wildcard(
     }
 }
 
+/// check whether the logical plan we are building is valid
+fn check_plan_invalid(plan: &LogicalPlan) -> Option<DataFusionError> {
+    match plan {
+        LogicalPlan::Projection(Projection { expr, input, .. })
+        | LogicalPlan::Sort(Sort { expr, input })
+        | LogicalPlan::Window(Window {
+            window_expr: expr,
+            input,
+            ..
+        }) => check_plan_invalid(input)
+            .or_else(|| check_any_invalid_expr(expr, input.schema())),
+
+        LogicalPlan::Filter(Filter {
+            predicate: expr,
+            input,
+        }) => {
+            check_plan_invalid(input).or_else(|| check_invalid_expr(expr, input.schema()))
+        }
+
+        LogicalPlan::Aggregate(Aggregate {
+            input,
+            group_expr,
+            aggr_expr,
+            ..
+        }) => {
+            let schema = input.schema();
+            check_plan_invalid(input)
+                .or_else(|| check_any_invalid_expr(group_expr, schema))
+                .or_else(|| check_any_invalid_expr(aggr_expr, schema))
+        }
+
+        LogicalPlan::Join(Join { left, right, .. })
+        | LogicalPlan::CrossJoin(CrossJoin { left, right, .. }) => {
+            check_plan_invalid(left).or_else(|| check_plan_invalid(right))
+        }
+
+        LogicalPlan::Repartition(Repartition { input, .. })
+        | LogicalPlan::Limit(Limit { input, .. })
+        | LogicalPlan::Explain(Explain { plan: input, .. })
+        | LogicalPlan::Analyze(Analyze { input, .. }) => check_plan_invalid(input),
+
+        LogicalPlan::Union(Union { inputs, .. }) => {
+            inputs.iter().filter_map(check_plan_invalid).next()
+        }
+
+        LogicalPlan::TableScan(TableScan {
+            table_name: _,
+            source,
+            projection,
+            projected_schema,
+            filters,
+            limit: _,
+        }) => {
+            if let Some(projection) = projection {
+                if projection.len() > projected_schema.fields().len() {
+                    return Some(DataFusionError::Plan(
+                        "projection contains columns that doesnt belong to projected schema"
+                            .to_owned(),
+                    ));
+                }
+            }
+            let schema = &source.schema().to_dfschema_ref().ok()?;
+            check_any_invalid_expr(filters, schema)
+        }
+
+        _ => None,
+    }
+}
+
+/// find first error in the exprs
+fn check_any_invalid_expr(
+    exprs: &[Expr],
+    schema: &DFSchemaRef,
+) -> Option<DataFusionError> {
+    exprs
+        .iter()
+        .filter_map(|e| check_invalid_expr(e, schema))
+        .next()
+}
```

Comment:
Hm, if use Result, code here looks like:
```
    for expr in exprs {
        let r = check_invalid_expr(expr, schema);
        if r.is_err() {
            return r;
        }
    }
    Ok(())
```
or
```
    exprs
        .iter()
        .map(|e| check_invalid_expr(e, schema))
        .fold_ok((), |start, v| start) // I dont like it
```
I dont like the `fold_ok((), |start, v| start)`
But now, it's 
```
    exprs
        .iter()
        .filter_map(|e| check_invalid_expr(e, schema))
        .next()
```
I don't know how to make it more elegant, please give more suggestions.
@jackwener @alamb

Reply:
how about:

```rust
for expr in exprs { 
    check_invalid_expr(expr, schema)?
}
Ok(())
```

You can also use https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_for_each

Reply:
Thanks!
I applied the suggested changes from @jackwener which is same as yours.